### PR TITLE
VRT_fail to be annoted with cold attribute as a codepath to be taken …

### PR DESCRIPTION
**Pull request by devnexen**
…less likely

### CODE REVIEW

**Review by nigoroll — APPROVED** [review]

**Comment by Dridi** [comment]
bugwash: generally OK but @bsdphk wants to check the effect on assembly code (same thing for #3020 that we want to reintroduce at some point).

File: include/vdef.h [hunk 1]
```diff
@@ -241,6 +241,12 @@ int __llvm_gcov_flush(void);
 #  define v_unused_
 #endif
 
+#ifdef __GNUC__
+#  define v_cold    __attribute__((cold))
+#else
+#  define v_cold
+#endif
```
Comment (Dridi) [inline include/vdef.h:248]: Should we use the appropriate `__GNUC_PREREQ__(x, y)` to check this one?

Comment (Dridi) [reply include/vdef.h:248]: Also, shouldn't it be `v_cold_` with a trailing `_` to match our other "vendor" macros?

**Comment by Dridi** [comment]
FWIW this one should have an effect on assembly at call sites, and an effect on linking.

**Comment by bsdphk** [comment]
I have not been able to actually measure any improvement, but inspection of the produced instructions show these calls booted down to a "suffix" after the return instruction, which, presumably, should make for fewer and more productive cache-lines for the "warm" cases.

The macro should be `v_cold_` to be consistent and the #ifdefery should be properly researched, if possible.

**Comment by devnexen** [comment]
> I have not been able to actually measure any improvement, but inspection of the produced instructions show these calls booted down to a "suffix" after the return instruction, which, presumably, should make for fewer and more productive cache-lines for the "warm" cases.
> 
> The macro should be `v_cold_` to be consistent 

ah my bad, I misred :)

and the #ifdefery should be properly researched, if possible.

well about the proper GCC version, the cold attribute seems to appear from the 4.3.x serie.

[4.2](https://gcc.gnu.org/onlinedocs/gcc-4.2.0/gcc/Function-Attributes.html) vs [4.3](https://gcc.gnu.org/onlinedocs/gcc-4.3.0/gcc/Function-Attributes.html)

**Comment by Dridi** [comment]
Summoning @fwsGonzo for an opinion.

**Comment by nigoroll** [comment]
@devnexen I just tried the gh rebase. Please feel free to discard that merge commit.

**Comment by Dridi** [comment]
Converted to a draft to not accidentally merge the merge commit.

**Comment by fwsGonzo** [comment]
LGTM. The cold attribute can be used on functions and labels, and makes the paths leading to it unlikely, which can create ideal prologues when you have "cold" assertions on entry. GCC may also put the cold parts into its own .text.cold section, however it's not necessary to get most of the benefits.

Suppose you have this function:
```
int hot_function()
{
    // hot path
    if (cond)
         cold_path();
    // resume hot path
}
```
Not taken branches are always faster than those that are taken, and since cold_path() is an error handling function the prediction unit will most likely remember the branch as always false. Still, we are at risk of the compiler inlining the function and losing icache to a failure block. The cold attribute solves it by making all of it unlikely, however it is a big hammer as it practically means "optimize for size", which may not always be what you want. For error handling paths it's probably fine. I suspect that just using __builtin_expect has the same effect, as an unlikely branch may disable function inlining.

There are a few cold functions in  glibc itself:
```
$ readelf -a ... | grep cold
    64: 0000000000401200   175 FUNC    LOCAL  DEFAULT    7 d_type.cold
   103: 000000000040142d    10 FUNC    LOCAL  DEFAULT    7 __gxx_personality_v0.cold
   108: 00000000004014f1    46 FUNC    LOCAL  DEFAULT    7 _Znwm.cold
   110: 000000000040151f    92 FUNC    LOCAL  DEFAULT    7 _ZnwmSt11align_val_t.cold
   156: 0000000000401c28     5 FUNC    LOCAL  DEFAULT    7 uw_install_context_1.cold
   158: 0000000000401c2d     5 FUNC    LOCAL  DEFAULT    7 read_encoded_value.cold
   160: 0000000000401c32    10 FUNC    LOCAL  DEFAULT    7 execute_stack_op.cold
   162: 0000000000401c3c    10 FUNC    LOCAL  DEFAULT    7 uw_update_context_1.cold
   164: 0000000000401c46     5 FUNC    LOCAL  DEFAULT    7 execute_cfa_program.cold
   166: 0000000000401c4b     5 FUNC    LOCAL  DEFAULT    7 uw_frame_state_for.cold
   169: 0000000000401c50     5 FUNC    LOCAL  DEFAULT    7 uw_init_context_1.cold
   174: 0000000000401c5f     5 FUNC    LOCAL  DEFAULT    7 _Unwind_GetGR.cold
   175: 0000000000401c64     5 FUNC    LOCAL  DEFAULT    7 _Unwind_SetGR.cold
   178: 0000000000401c6e     5 FUNC    LOCAL  DEFAULT    7 _Unwind_Resume.cold
   180: 0000000000401c78     5 FUNC    LOCAL  DEFAULT    7 _Unwind_Backtrace.cold
   195: 0000000000401c91     5 FUNC    LOCAL  DEFAULT    7 add_fdes.cold
   197: 0000000000401c96     5 FUNC    LOCAL  DEFAULT    7 linear_search_fdes.cold
   207: 0000000000401ca5     5 FUNC    LOCAL  DEFAULT    7 search_object.cold
   212: 0000000000401caa     5 FUNC    LOCAL  DEFAULT    7 _Unwind_Find_FDE.cold
   215: 0000000000401caf    15 FUNC    LOCAL  DEFAULT    7 __assert_fail_base.cold
   218: 0000000000401cbe     5 FUNC    LOCAL  DEFAULT    7 _nl_load_domain.cold
   223: 0000000000401ef7    87 FUNC    LOCAL  DEFAULT    7 _IO_new_fclose.cold
   225: 0000000000401f4e    86 FUNC    LOCAL  DEFAULT    7 _IO_fflush.cold
   227: 0000000000401fa4    86 FUNC    LOCAL  DEFAULT    7 _IO_fputs.cold
   229: 0000000000401ffa    86 FUNC    LOCAL  DEFAULT    7 _IO_fread.cold
   231: 0000000000402050    86 FUNC    LOCAL  DEFAULT    7 _IO_ftell.cold
   233: 00000000004020a6    86 FUNC    LOCAL  DEFAULT    7 _IO_fwrite.cold
   235: 00000000004020fc    87 FUNC    LOCAL  DEFAULT    7 _IO_puts.cold
   237: 0000000000402153     5 FUNC    LOCAL  DEFAULT    7 _IO_seekoff_unlocked.cold
   238: 0000000000402158    91 FUNC    LOCAL  DEFAULT    7 _IO_seekoff.cold
   241: 00000000004021b3    87 FUNC    LOCAL  DEFAULT    7 _IO_wfile_underflow.cold
   246: 000000000040220a    87 FUNC    LOCAL  DEFAULT    7 fputc.cold
   248: 0000000000402261    86 FUNC    LOCAL  DEFAULT    7 fseek.cold
   250: 00000000004022b7    87 FUNC    LOCAL  DEFAULT    7 getchar.cold
   262: 0000000000402373     5 FUNC    LOCAL  DEFAULT    7 round_and_return.cold
   269: 0000000000402378     5 FUNC    LOCAL  DEFAULT    7 round_and_return.cold
   276: 000000000040237d     5 FUNC    LOCAL  DEFAULT    7 round_and_return.cold
   283: 0000000000402382     5 FUNC    LOCAL  DEFAULT    7 round_and_return.cold
   289: 0000000000402387     5 FUNC    LOCAL  DEFAULT    7 __printf_fp_l.cold
   292: 000000000040238c     5 FUNC    LOCAL  DEFAULT    7 __printf_fphex.cold
   294: 0000000000402391    86 FUNC    LOCAL  DEFAULT    7 _IO_getdelim.cold
   300: 00000000004023f1     5 FUNC    LOCAL  DEFAULT    7 __gcc_personality_v0.cold
```

**Comment by nigoroll** [comment]
bugwash:

* phk and me think that this is unlikely to cause harm
* but we should be careful to only apply it for "truly abnormal paths, asserts and probably also VRT_fail" (quote phk)

**Pull request merged**